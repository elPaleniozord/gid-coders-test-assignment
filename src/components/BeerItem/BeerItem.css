.item-wrapper {
  border-radius: 3px;
  margin: .4rem;
  padding: .3rem;
  padding-bottom: 1rem;
  background: var(--offwhite);
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  height: auto;
  cursor: pointer;
  max-width: 700px;
}
.item-header {
  color: var(--offblack);
  font-size: 2.3rem;
  text-transform: uppercase;
  margin: 0;
  max-width: calc(100% - 70px);
  text-align: left;
}
.item-tagline {
  margin: 0;
  text-align: left;
}

.item-info {
  display: flex;
  width: 150px;
  justify-content: space-between;
}
.item-info > * {
  margin-top: 5px;
}

.item-img {
  height: 200px;
  margin-left: 10px;
}

.item-details {
  transition: all .4s ease-in-out;
  max-height: 0;
  height: 100%;
  overflow: hidden;
  opacity: 0;
}
.item-details__open {
  max-height: 1200px;
  opacity: 1;
}

.item-button {
  position: absolute;
  right: .3rem;
}

.item-button__fav {
  border: none;
  background: none;
}

.fav-icon {
  height: 44px;
  width: 44px;
  transform: rotate(15deg);
  cursor: pointer;
}
.fav-icon > path {
  transition: all .3s ease-in-out;
}

.item-button__expand {
  bottom: 15px;
}
.details-description {
  display: flex;
  align-items: center;
}
.details-color {
  display: flex;
  align-items: center;
}

.details-color__indicator {
  display: block;
  background: var(--highlight);
  width: 3px;
  height: 28px;
}

.details-color__body {
  margin-left: 20px;
  border-radius: 11px;
  height: 22px;
  width: 200px;
  background: linear-gradient(90deg, rgba(255,255,255,1) 0%, rgba(249,239,111,1) 0%, rgba(255,126,0,1) 50%, rgba(0,0,0,1) 100%);
}

.details-food {
  text-align: left;
}