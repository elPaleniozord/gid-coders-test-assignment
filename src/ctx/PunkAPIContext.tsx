import React, { useEffect, useReducer } from "react"
import { ActionType, punkAPIReducer } from "./PunkAPIReducer"

export interface Beer {
  id: number
  [key: string]: string | number | boolean
}

interface APIContext {
  loading: boolean
  beers: Beer[]
  favorites: number[]
  fetchMore?: () => void
  handleFav?: (beer:Beer) => void
  // dispatch: () => null
}

const defaults = {
  loading: false,
  beers: [],
  favorites: []
}

export const PunkAPIContext = React.createContext<APIContext>(defaults)

const initialState = {
  beers: [],
  loading: false,
  page: 1,
  favorites: []
}

export const PunkAPIProvider: React.FC = ({children}) => {
  const [state, dispatch] = useReducer(punkAPIReducer, initialState)

  const handleLocalStorage: ()=> Beer[] = () => {
      const stored = localStorage.getItem('favorite-beers')
      return stored ? JSON.parse(stored) : []
  }

  const handleFav = async (beer: Beer) => {
    const idx = state.favorites.indexOf(beer.id)
    const update = handleLocalStorage()

    if(idx === -1) update.push(beer)
    else update.splice(idx,1)
    
    localStorage.setItem('favorite-beers', JSON.stringify(update))
    const ids = update.map(beer => beer.id)

    dispatch({type: ActionType.UpdateFavorites, payload: ids})
  }

  useEffect(() => {
    const favs = handleLocalStorage().map(beer => beer.id)
    dispatch({type: ActionType.ToggleLoading})
    fetch(`https://api.punkapi.com/v2/beers?page=${state.page}&per_page=${40}`)
      .then(res => {
        if(res.ok) return res.json()
        throw res
      })
      .then(data => {
        dispatch(
          state.page === 1 ? 
          { type: ActionType.SetBeers, payload: [data, favs] }
          :
          { type: ActionType.UpdateBeers, payload: data }
        )
      })
      .catch(err => console.log(err))
  }, [state.page])

  return (
    <PunkAPIContext.Provider 
      value={{
        loading: state.loading,
        beers: state.beers,
        favorites: state.favorites,
        fetchMore: () => dispatch({type: ActionType.FetchNextPage}),
        handleFav: handleFav,
      }}
    >
      {children}
    </PunkAPIContext.Provider>
  )
}