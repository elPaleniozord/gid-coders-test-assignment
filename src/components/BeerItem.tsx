import { useContext, useState } from 'react'
import { Beer, PunkAPIContext } from '../ctx/PunkAPIContext'

const BeerItem: React.FC<Beer> = (props) => {
  const [expanded, setExpanded] = useState(false)
  const {favorites, handleFav} = useContext(PunkAPIContext)

  return (
    <li>
      <h2>{props.name}</h2>
      <div>{props.description}</div>
      <button onClick={() => handleFav && handleFav(props)}>
        {favorites.includes(props.id) ? 'Remove from favorites' : 'Add to favorites'}
      </button>
      <button onClick={() => setExpanded(!expanded)}>{expanded ? `Collapse` : `Expand`}</button>
    </li>
  )
}

export default BeerItem