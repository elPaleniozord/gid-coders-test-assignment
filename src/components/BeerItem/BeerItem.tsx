import { MouseEventHandler, useContext, useState } from 'react'
import { Beer, PunkAPIContext } from '../../ctx/PunkAPIContext'
import Ingredients from '../Ingredients/Ingredients'
import './BeerItem.css'

const BeerItem: React.FC<Beer> = (props) => {
  const [expanded, setExpanded] = useState(false)
  const {favorites, handleFav} = useContext(PunkAPIContext)

  //calculates indicator offset
  const colorScale = (ebc: number) => {
    const range = [4,6,8,12,16,20,26,33,39,47,57,69,79]
    const width = 200
    let idx = range.findIndex(val => ebc < val)
    if(idx === -1) idx = 11.5
    return width - ((200/12) * idx)
  }

  const handleFavorite: MouseEventHandler = (e) => {
    e.stopPropagation()
    handleFav && handleFav(props)
  }

  const handleExpand: MouseEventHandler = (e) => {
    e.stopPropagation()
    setExpanded(!expanded)
  }

  return (
    <li className='item-wrapper' onClick={handleExpand}>
      <h2 className='item-header'>{props.name}</h2>
      <h3 className='item-tagline'>{props.tagline}</h3>
      <div className='item-info'>
        <p className='item-info__abv'>ABV: {props.abv}%</p>
        <p className='item-info__ibu'>IBU: {props.ibu}%</p>
      </div>

      <button className='item-button item-button__fav' onClick={handleFavorite}>
        <svg className='fav-icon' xmlns="http://www.w3.org/2000/svg" viewBox="0 0 260 245">
          <path d="m56,237 74-228 74,228L10,96h240" fill={favorites.includes(props.id) ? '#00e1ff' : '#829094'}/>
        </svg>
      </button>

      <div className={`item-details ${expanded ? 'item-details__open': ''}`}>
        <div className='details-description'>
          <p>{props.description}</p>
          <img className='item-img' src={props.image_url} alt={`${props.name}`} />
        </div>
        <div className='details-color'>
          <h3>Color:</h3>
          <div className='details-color__body'></div>
          <div className='details-color__indicator' style={{transform: `translateX(-${colorScale(props.ebc)}px)`}}></div>
        </div>
        <Ingredients hops={props.ingredients.hops} malts={props.ingredients.malt} yeast={props.ingredients.yeast} />

        <div className='details-food'>
          <h3>Best served with:</h3>
          <ul>{props.food_pairing.map((food: string, i:number) => (<li key={`fp-${i}`}>{food}</li>))}</ul>
        </div>
      </div>
    </li>
  )
}

export default BeerItem