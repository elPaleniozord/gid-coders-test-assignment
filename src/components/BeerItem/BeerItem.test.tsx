import { render } from '@testing-library/react';
import { sampleBeer } from '../../ctx/PunkAPIContext.test';
import BeerItem from "./BeerItem"

test('should render Favorites Component properly', () => {
  const { container } = render(<BeerItem {...sampleBeer} />)
  const favorites = container.querySelector('li')
  expect(favorites).toMatchSnapshot()
})