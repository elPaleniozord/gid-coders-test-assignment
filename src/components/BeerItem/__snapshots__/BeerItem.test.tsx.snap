// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Favorites Component properly 1`] = `
<li
  class="item-wrapper"
>
  <h2
    class="item-header"
  >
    Punk IPA 2007 - 2010
  </h2>
  <h3
    class="item-tagline"
  >
    Post Modern Classic. Spiky. Tropical. Hoppy.
  </h3>
  <div
    class="item-info"
  >
    <p
      class="item-info__abv"
    >
      ABV: 
      6
      %
    </p>
    <p
      class="item-info__ibu"
    >
      IBU: 
      60
      %
    </p>
  </div>
  <button
    class="item-button item-button__fav"
  >
    <svg
      class="fav-icon"
      viewBox="0 0 260 245"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m56,237 74-228 74,228L10,96h240"
        fill="#829094"
      />
    </svg>
  </button>
  <div
    class="item-details "
  >
    <div
      class="details-description"
    >
      <p>
        Our flagship beer that kick started the craft beer revolution. This is James and Martin's original take on an American IPA, subverted with punchy New Zealand hops. Layered with new world hops to create an all-out riot of grapefruit, pineapple and lychee before a spiky, mouth-puckering bitter finish.
      </p>
      <img
        alt="Punk IPA 2007 - 2010"
        class="item-img"
        src="https://images.punkapi.com/v2/192.png"
      />
    </div>
    <div
      class="details-color"
    >
      <h3>
        Color:
      </h3>
      <div
        class="details-color__body"
      />
      <div
        class="details-color__indicator"
        style="transform: translateX(-116.66666666666666px);"
      />
    </div>
    <div
      class="ingredients-container"
    >
      <h3
        class="ingredients-header"
      >
        Ingredients
      </h3>
      <div
        class="ingredients-group"
      >
        <img
          alt="hops-icon"
          class="ingredients-icon"
          src="/hops-icon.png"
        />
        <ul
          class="ingredients-list"
        >
          <li
            class="ingredients-item"
          >
            Ahtanum
          </li>
          <li
            class="ingredients-item"
          >
            Chinook
          </li>
        </ul>
      </div>
      <div
        class="ingredients-group"
      >
        <img
          alt="malt-icon"
          class="ingredients-icon"
          src="/malt-icon.png"
        />
        <ul
          class="ingredients-list"
        >
          <li
            class="ingredients-item"
          >
            Extra Pale
          </li>
        </ul>
      </div>
      <div
        class="ingredients-group"
      >
        <img
          alt="yeast-icon"
          class="ingredients-icon"
          src="/yeast-icon.png"
        />
        <ul
          class="ingredients-list"
        >
          <li
            class="ingredients-item"
          >
            Wyeast 1056 - American Aleâ„¢
          </li>
        </ul>
      </div>
    </div>
    <div
      class="details-food"
    >
      <h3>
        Best served with:
      </h3>
      <ul>
        <li>
          Spicy carne asada with a pico de gallo sauce
        </li>
        <li>
          Shredded chicken tacos with a mango chilli lime salsa
        </li>
        <li>
          Cheesecake with a passion fruit swirl sauce
        </li>
      </ul>
    </div>
  </div>
</li>
`;
